sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -163,8 +163,35 @@ Scheduler::GraphBuilder::getOrInsertMemObjRecord(const QueueImplPtr &Queue,
     return Record;
 
   const size_t LeafLimit = 8;
-  MemObject->MRecord.reset(
-      new MemObjRecord{Queue->getContextImplPtr(), LeafLimit});
+  CircularBufferExtended::IfGenericIsFullF IfGenericIsFull =
+      [this](Command *Cmd, MemObjRecord *Record,
+             CircularBuffer<Command *> &Leaves) {
+        Command *OldLeaf = Leaves.front();
+        // TODO this is a workaround for duplicate leaves, remove once fixed
+        if (OldLeaf == Cmd)
+          return;
+        // Add the old leaf as a dependency for the new one by duplicating one
+        // of the requirements for the current record
+        DepDesc Dep = findDepForRecord(Cmd, Record);
+        Dep.MDepCommand = OldLeaf;
+        Cmd->addDep(Dep);
+        OldLeaf->addUser(Cmd);
+        --(OldLeaf->MLeafCounter);
+      };
+  CircularBufferExtended::AllocateDependencyF AllocateDependency =
```

Comment:
Can we get rid of `IfGenericIsFull` and get by with passing just `AllocateDependency` to `CircularBufferExtended`? The 2 lambdas are mostly identical and we can move the checks from `IfGenericIsFull` to `CircularBufferExtended`.

Reply:
Nice catch.
Done.